src/backend/access/appendonly/appendonlywriter.c

```diff
@@ -1641,6 +1683,38 @@ UpdateMasterAosegTotalsFromSegments(Relation parentrel,
 		}
 	}
 
+	if (appendonly_compaction_segno != NIL)
+	{
+		AORelHashEntryData *aoentry;
+
+		acquire_lightweight_lock();
+
+		aoentry = AORelGetOrCreateHashEntry(RelationGetRelid(parentrel));
+		Assert(aoentry);
+
+		foreach(l, appendonly_compaction_segno)
+		{
+			int			qe_segno = lfirst_int(l);
+
+			Assert(qe_segno >= 0);
+
+			if (!awaiting_drop[qe_segno])
+			{
+				elogif(Debug_appendonly_print_segfile_choice, LOG,
+					   "Update the file segment state after compaction: "
+					   "relation \"%s\" (%d), segno %d",
+					   get_rel_name(RelationGetRelid(parentrel)),
+					   RelationGetRelid(parentrel),
+					   qe_segno);
+
+				aoentry->relsegfiles[qe_segno].state = DROP_USE;
```

Comment:
@RekGRpth, @andr-sokolov, @dnskvlnk, WHY is segment marked as DROP_USE, if compaction didn't happen? Are there any cases when this state will stay here and data will be unavailable?

Reply:
> WHY is segment marked as DROP_USE

Because the state machine at the end of the transaction will scroll the state to the default value.

Reply:
> Because the state machine at the end of the transaction will scroll the state to the default value.

in case of commit only. What about abort? Anyway, it's very dirty hack and may be unexpected for further developers.

Reply:
I can use INSERT_USE, then the state machine will scroll the state to the default value regardless of the commit or abort.

Reply:
Can you explain, why INSERT_USE? After your modifications, vacuum doesn't touch given AO-segment at all. What about sideeffects of this state?

Reply:
I changed it to the default value and added its processing to the state machine.

Reply:
Why do we need additional actions for default state? Do we really need them?

Reply:
Yes, you are right, I changed it.